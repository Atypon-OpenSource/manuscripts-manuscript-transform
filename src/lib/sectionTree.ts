/*!
 * Â© 2019 Atypon Systems LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Model } from '@manuscripts/manuscripts-json-schema'

interface SectionProperties {
  path?: string[]
  priority?: number
}

export default (modelMap: Map<string, Model & SectionProperties>) => ({
  getPath(id: string) {
    const doc = modelMap.get(id)
    if (!doc) {
      throw new Error(`Doc ${id} not found`)
    }
    const path = doc.path
    if (!path) {
      throw new Error(`Doc ${id} is not a Section`)
    }

    return path
  },

  parent(id: string) {
    const path = this.getPath(id)
    const parentId = path[path.length - 2]
    if (!parentId) {
      return null
    }
    return modelMap.get(parentId) || null
  },

  children(id: string | null) {
    const children: Array<Model & SectionProperties> = []
    modelMap.forEach(model => {
      let path
      try {
        path = this.getPath(model._id)
      } catch (e) {
        return
      }

      if (id === null && path.length === 1 && path[0] === model._id) {
        children.push(model)
      }

      if (path[path.length - 2] === id) {
        children.push(model)
      }
    })

    return children.sort((a, b) => a.priority! - b.priority!)
  },

  siblings(id: string) {
    const parent = this.parent(id)
    return this.children(parent && parent._id)
  },
})
