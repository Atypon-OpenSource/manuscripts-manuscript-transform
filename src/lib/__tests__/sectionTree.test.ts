/*!
 * Â© 2019 Atypon Systems LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import projectDump3 from '@manuscripts/examples/data/project-dump-3.json'
import { Model } from '@manuscripts/manuscripts-json-schema'
import sectionTree from '../sectionTree'
const deepFreeze = require('deep-freeze') // tslint:disable-line:no-var-requires

interface ContainedModel extends Model {
  containerID: string
  manuscriptID?: string
  // [key: string]: any
}

const withRequiredProperties = <T extends ContainedModel>(
  model: Partial<T>
): T =>
  ({
    createdAt: 0,
    updatedAt: 0,
    containerID: 'MPProject:project-1',
    manuscriptID: 'MPManuscript:manuscript-1',
    ...model,
  } as T) // tslint:disable-line:no-object-literal-type-assertion

const data: ContainedModel[] = [
  {
    objectType: 'MPSection',
    _id: 'MPSection:top-level-1',
    priority: 1,
    path: ['MPSection:top-level-1'],
  },
  {
    objectType: 'MPSection',
    _id: 'MPSection:top-level-2',
    priority: 2,
    path: ['MPSection:top-level-2'],
  },
  {
    objectType: 'MPSection',
    _id: 'MPSection:child-2',
    priority: 2,
    path: ['MPSection:top-level-2', 'MPSection:child-2'],
  },
  {
    objectType: 'MPSection',
    _id: 'MPSection:child-1',
    priority: 1,
    path: ['MPSection:top-level-2', 'MPSection:child-1'],
  },
  {
    objectType: 'MPCitation',
    _id: 'MPCitation:not-a-section',
  },
].map(withRequiredProperties)

const modelMap = new Map(
  data.map(model => {
    return [model._id, model]
  })
)
deepFreeze(modelMap)

describe('sectionTree', () => {
  describe('parent', () => {
    it('should find a sections parent', () => {
      const result = sectionTree(modelMap).parent('MPSection:child-1')
      expect(result).toHaveProperty('_id', 'MPSection:top-level-2')
    })

    it('should return null for top-level parents', () => {
      const result = sectionTree(modelMap).parent('MPSection:top-level-2')
      expect(result).toBeNull()
    })

    it('should throw if called with a non-section', () => {
      expect(() => {
        sectionTree(modelMap).parent('MPCitation:not-a-section')
      }).toThrow()
    })
  })

  describe('children', () => {
    it('should find all children in order', () => {
      const result = sectionTree(modelMap).children('MPSection:top-level-2')
      expect(result).toHaveLength(2)
      expect(result[0]).toHaveProperty('_id', 'MPSection:child-1')
      expect(result[1]).toHaveProperty('_id', 'MPSection:child-2')
    })
  })

  describe('siblings', () => {
    it('should find a sections siblings in order', () => {
      const result = sectionTree(modelMap).siblings('MPSection:child-1')
      expect(result).toHaveLength(2)
      expect(result[0]).toHaveProperty('_id', 'MPSection:child-1')
      expect(result[1]).toHaveProperty('_id', 'MPSection:child-2')
    })

    it('should work for the top-level sections', () => {
      const result = sectionTree(modelMap).siblings('MPSection:top-level-1')
      expect(result).toHaveLength(2)
      expect(result[0]).toHaveProperty('_id', 'MPSection:top-level-1')
      expect(result[1]).toHaveProperty('_id', 'MPSection:top-level-2')
    })

    it('should throw if called on a non-section', () => {
      expect(() => {
        sectionTree(modelMap).siblings('MPCitation.not-a-section')
      }).toThrow()
    })

    it('should match snapshot', () => {
      /* tslint:disable-next-line:no-any */
      const modelMap: any = new Map(
        projectDump3.data.map(model => {
          return [model._id, model]
        })
      )
      deepFreeze(modelMap)
      const result = sectionTree(modelMap).siblings(
        'MPSection:786116A4-364C-428B-9B05-FE447B1AEEFB'
      )
      expect(result).toMatchSnapshot()
    })
  })
})
