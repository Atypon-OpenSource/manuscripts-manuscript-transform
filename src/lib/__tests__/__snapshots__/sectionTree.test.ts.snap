// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sectionTree siblings should match snapshot 1`] = `
Array [
  Object {
    "_id": "MPSection:786116A4-364C-428B-9B05-FE447B1AEEFB",
    "containerID": "MPProject:ADC69637-C321-4158-8E24-F92068D4727D",
    "createdAt": 1558720146,
    "elementIDs": Array [],
    "manuscriptID": "MPManuscript:B3BB2CD8-F944-47C3-9F01-1996DBD417EE",
    "objectType": "MPSection",
    "path": Array [
      "MPSection:369BC11C-4281-4BED-9AA3-5ABE066BF575",
      "MPSection:786116A4-364C-428B-9B05-FE447B1AEEFB",
    ],
    "priority": 2,
    "sessionID": "02b8ed9b-43bc-41a0-9732-f478f7fab0cc",
    "title": "Subsection",
    "updatedAt": 1558720168,
  },
]
`;
